In [34]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from scipy import stats

from tabula import read_pdf

pdf_url = 'https://globalassets.starbucks.com/assets/94fbcc2ab1e24359850fa1870fc988bc.pdf'

data = read_pdf(pdf_url, pages='all',output_format='dataframe',
                area=[76.87,72.73,531.76,657.89], guess=False, pandas_options={'header':None})

data.head()

0       1    2    3       4    5    6    7    8   \
0                         NaN     NaN  NaN  NaN     NaN  NaN  NaN  NaN  NaN   
1                      Coffee     NaN  NaN  NaN     NaN  NaN  NaN  NaN  NaN   
2  Brewed Coffee - Dark Roast   Short  NaN  NaN  236 mL  3.0  0.1  0.0  0.0   
3  Brewed Coffee - Dark Roast    Tall  NaN  NaN  354 mL  4.0  0.1  0.0  0.0   
4  Brewed Coffee - Dark Roast  Grande  NaN  NaN  473 mL  5.0  0.1  0.0  0.0   

    9    10   11   12   13   14   15   16   17   18   19  
0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2  0.0    5    0    0  0.0  0.3   0%   0%   0%   0%  130  
3  0.0   10    0    0  0.0  0.5   0%   0%   0%   0%  193  
4  0.0   10    0    0  0.0    1   0%   0%   0%   0%  260

In [36]:
data.shape

(1284, 20)

# The following columns are represented in 'grams'
- Total Fat
- Saturated Fat
- Trans Fat
- Total Carbs
- Dietary Fiber
- Sugar
- Protein

# The following columns are represented in 'milligrams'
- Cholesterol
- Sodium
- Caffeine

In [51]:
columns={
    0: 'Product Name',
    1: 'Size',
    2: 'Milk',
    3: 'Whip',
    4: 'Serving Size',
    5: 'Calories',
    6: 'Total Fat',
    7: 'Saturated Fat',
    8: 'Trans Fat',
    9: 'Cholesterol',
    10: 'Sodium',
    11: 'Total Carbs',
    12: 'Dietary Fiber',
    13: 'Sugar',
    14: 'Protein',
    15: 'Vitamin A',
    16: 'Vitamin C',
    17: 'Calcium',
    18: 'Iron',
    19: 'Caffeine'
}
starbucks = data.copy()
starbucks = starbucks.rename(columns=columns)

starbucks.head()

Product Name    Size Milk Whip Serving Size  Calories  \
0                         NaN     NaN  NaN  NaN          NaN       NaN   
1                      Coffee     NaN  NaN  NaN          NaN       NaN   
2  Brewed Coffee - Dark Roast   Short  NaN  NaN       236 mL       3.0   
3  Brewed Coffee - Dark Roast    Tall  NaN  NaN       354 mL       4.0   
4  Brewed Coffee - Dark Roast  Grande  NaN  NaN       473 mL       5.0   

   Total Fat  Saturated Fat  Trans Fat  Cholesterol Sodium Total Carbs  \
0        NaN            NaN        NaN          NaN    NaN         NaN   
1        NaN            NaN        NaN          NaN    NaN         NaN   
2        0.1            0.0        0.0          0.0      5           0   
3        0.1            0.0        0.0          0.0     10           0   
4        0.1            0.0        0.0          0.0     10           0   

  Dietary Fiber  Sugar Protein Vitamin A Vitamin C Calcium Iron Caffeine  
0           NaN    NaN     NaN       NaN       NaN     NaN  NaN      NaN  
1           NaN    NaN     NaN       NaN       NaN     NaN  NaN      NaN  
2             0    0.0     0.3        0%        0%      0%   0%      130  
3             0    0.0     0.5        0%        0%      0%   0%      193  
4             0    0.0       1        0%        0%      0%   0%      260

In [53]:
starbucks.isna().sum().sort_values(ascending=False)

Whip             680
Milk             214
Serving Size      65
Caffeine          41
Product Name      30
Size              25
Protein           20
Sugar             20
Dietary Fiber     20
Total Carbs       20
Sodium            20
Iron              20
Trans Fat         20
Saturated Fat     20
Total Fat         20
Calories          20
Vitamin A         20
Vitamin C         20
Calcium           20
Cholesterol       20
dtype: int64

# Inference about NaNs

- **Whip** has 680 NaNs because whipped cream isn't an add-on for those items
- **Milk** has 214 NaNs because milk can't be added to those items
- **Serving Size** The NaNs come from categorical grouping of *drinks that have an emphasis on espresso, or they are just addons.
- **Caffeine** These items just don't have caffeine
- 

## TODOs:
**1)** Figure out the NaNs / why there are NaNs  
**2)** Product Name column has 66 NaNs, where it should be the drink size  
**3)** Serving size is a string, needs to be an integer/float preferably float  
**4)** Milk is also a string with % signs in it  
**5)** Actually a lot of columns have % signs in them so we'll have to do something about that  
**6)** Remove the ROWS that categorize the drinks, and instead create a new column that categorizes them  
**7)** Also rename those columns
